design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,124 @@
+  - `platform/darwin` or `platform/apple`
+  - `platform/windows`
+  - `platform/qt`
+- `test`: common (headless) tests for all platforms; disabled only in exceptional cases if some features are not applicable due to the nature of a specific platform
+  - only platform-independent tests should go here
+  - platfom-specific tests should go under `platform/<platform>/test`
+
+The core itself ideally should have no binaries unless they are really platform-independent (e.g. headless rendering).
+
+### Supported platforms
+
+We can split the platforms into two categories:
+
+- **Native platforms / operating systems**
+  - Windows
+  - Linux
+  - Darwin and its derivatives (macOS, iOS, tvOS, ...)
+  - Android
+- **Cross-platform engines**
+  - Node.js
+  - Qt
+  - GLFW
+
+### Build infrastructure
+
+The core should be exclusively built by `CMake`. All currently supported platforms are supported by it. No wrapper script or `Makefile` should be used. Can export `Xcode` projects if needed.
+
+The functionality part of the library (under `src`) should yield a separate object library. Similarly the platform-specific part (under `platform`) should be separate. Each 3rd-party dependency should also build as its own object library if distributed together with MapLibre. Static libraries may only be used as backwards compatibility for older platforms that the community may be interested in.
+
+The code should be C++17 or later (unless required otherwise by platform-specific integration).
+
+Compilation flags should ideally be the same on one native platform / operating system. This would allow building multiple platform-specific parts in parallel and would speed-up the development.
+
+Final products of the core would be two or more object libraries to be directly used by the SDKs or one static or dynamic library for binary distribution. *Note: At this point it is not clear if all platforms need binary distribution independent of the SDK that will be provided for that platform.*
+
+Examples of binary distributions could be:
+
+- Dynamic libraries for Windows, Linux and Android.
+- Dynamic and static frameworks for macOS and iOS.
+- Cross-platform engines will probably build the core as part of the SDK and no direct core binaries are needed (as written above this may also apply for native platforms).
+
+### Testing
+
+All testing should be done for all platforms at the same time and as complete as possible. Basic rendering tests should be done headlessly for all platforms.
+
+If needed for specific platforms (iOS comes to mind), a dedicated test app could be made to run the necessary rendering tests. This app should implement the minimal functionality that does not include the actual full-fledged integration of MapLibre (e.g. gestures, widgets, ...).
+
+Testing should be done by `CTest` if the platform allows (may not be possible for all integration tests on mobile devices).
```

Comment:
There is no reason why we should require platforms to wrap their own testing framework with CTest.

Reply:
This is for core only.